models/issue_dependency.go

```diff
@@ -0,0 +1,137 @@
+type IssueDependency struct {
+	ID           int64     `xorm:"pk autoincr"`
+	UserID       int64     `xorm:"NOT NULL"`
+	IssueID      int64     `xorm:"NOT NULL"`
+	DependencyID int64     `xorm:"NOT NULL"`
+	Created      time.Time `xorm:"-"`
+	CreatedUnix  int64     `xorm:"INDEX created"`
+	Updated      time.Time `xorm:"-"`
+	UpdatedUnix  int64     `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (exists bool, err error) {
+	sess := x.NewSession()
+
+	// Check if it aleready exists
+	exists, err = issueDepExists(x, issue.ID, dep.ID)
+	if err != nil {
+		return exists, err
+	}
+
+	// If it not exists, create it, otherwise show an error message
+	if !exists {
+		newIssueDependency := &IssueDependency{
+			UserID:       user.ID,
+			IssueID:      issue.ID,
+			DependencyID: dep.ID,
+		}
+
+		if _, err := x.Insert(newIssueDependency); err != nil {
+			return exists, err
+		}
+
+		// Add comment referencing the new dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+			return exists, err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+			return exists, err
+		}
+	}
+	return exists, nil
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+
+	// Check if it exists
+	var exists bool
+	if exists, err = issueDepExists(x, issue.ID, dep.ID); err != nil {
+		return err
+	}
+
+	// If it exists, remove it, otherwise show an error message
+	if exists {
+
+		var issueDepToDelete IssueDependency
+
+		switch depType {
+		case DependencyTypeBlockedBy:
+			issueDepToDelete = IssueDependency{IssueID: issue.ID, DependencyID: dep.ID}
+		case DependencyTypeBlocking:
+			issueDepToDelete = IssueDependency{IssueID: dep.ID, DependencyID: issue.ID}
+		default:
+			return
+		}
+
+		if _, err := x.Delete(&issueDepToDelete); err != nil {
+			return err
+		}
+
+		// Add comment referencing the removed dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, false); err != nil {
+			return err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (exists bool, err error) {
+
+	exists, err = e.Where("(issue_id = ? AND dependency_id = ?) OR (issue_id = ? AND dependency_id = ?)", issueID, depID, depID, issueID).Exist(&IssueDependency{})
```

Comment:
Shouldn't this be 2 different errors? `badf00d is already a dependency of 600df00d` vs `can not create circular dependencies` 🤔

Reply:
You're right. It simply was easier to resolve it that way 😁

Reply:
Just added it, hope I'm doing it right.